File: framework/rules/java/secure-coding.java
// Rule fixtures for framework/rules/java/secure-coding.yml
//
// `ruleid:` asserts the rule must fire on the next line; `ok:` asserts it must
// not. The negative cases are the PreparedStatement, ProcessBuilder and
// correctly-configured forms a reviewer would ask for.

import java.io.PrintWriter;
import java.security.MessageDigest;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.Statement;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import javax.servlet.http.HttpServletResponse;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.slf4j.Logger;
import org.springframework.http.ResponseEntity;

public class SecureCodingSamples {

    private Logger logger;

    public ResponseEntity<String> exceptionMessageReturned() {
        try {
            risky();
        } catch (Exception e) {
            // ruleid: devsecops-framework.secure-coding.java.exception-message-to-response
            return ResponseEntity.badRequest().body(e.getMessage());
        }
        return ResponseEntity.ok("done");
    }

    public ResponseEntity<String> exceptionLoggedNotReturned() {
        try {
            risky();
        } catch (Exception e) {
            // Correct handling: logged server-side, generic message returned.
            // ok: devsecops-framework.secure-coding.java.exception-message-to-response
            logger.error("operation failed", e);
            return ResponseEntity.status(500).body("An internal error occurred.");
        }
        return ResponseEntity.ok("done");
    }

    public void stackTraceCases(HttpServletResponse resp) throws Exception {
        try {
            risky();
        } catch (Exception e) {
            // ruleid: devsecops-framework.secure-coding.java.stack-trace-to-response
            e.printStackTrace(resp.getWriter());
        }

        try {
            risky();
        } catch (Exception e) {
            // ok: devsecops-framework.secure-coding.java.stack-trace-to-response
            logger.error("operation failed", e);
        }
    }

    public void sqlCases(Connection conn, String userId) throws Exception {
        Statement stmt = conn.createStatement();
        // ruleid: devsecops-framework.secure-coding.java.sql-string-concatenation
        stmt.executeQuery("SELECT * FROM users WHERE id = " + userId);

        // ok: devsecops-framework.secure-coding.java.sql-string-concatenation
        PreparedStatement ps = conn.prepareStatement("SELECT * FROM users WHERE id = ?");
        ps.setString(1, userId);
    }

    public void commandCases(String host) throws Exception {
        // ruleid: devsecops-framework.secure-coding.java.command-execution
        Runtime.getRuntime().exec("ping -c 1 " + host);

        // ok: devsecops-framework.secure-coding.java.command-execution
        new ProcessBuilder("ping", "-c", "1", host).start();
    }

    public void deserializationCases(java.io.InputStream in) throws Exception {
        // ruleid: devsecops-framework.secure-coding.java.insecure-deserialization
        Object obj = new java.io.ObjectInputStream(in).readObject();

        // ok: devsecops-framework.secure-coding.java.insecure-deserialization
        String json = new String(in.readAllBytes());
    }

    public void hashCases() throws Exception {
        // ruleid: devsecops-framework.secure-coding.java.weak-password-hash
        MessageDigest weak = MessageDigest.getInstance("MD5");

        // ok: devsecops-framework.secure-coding.java.weak-password-hash
        MessageDigest strong = MessageDigest.getInstance("SHA-256");
    }

    private void risky() throws Exception { }
}

// The type is written with the short name, as real code does after importing it.
// A fully-qualified `java.security.cert.X509Certificate[]` is a different AST
// node from the pattern's `X509Certificate[]`, so the earlier fixture never
// exercised this rule at all -- the fixture was wrong, not the rule.
class TrustManagerSamples {
    // ruleid: devsecops-framework.secure-coding.java.trust-all-certificates
    public void checkServerTrusted(X509Certificate[] chain, String authType) { }

    // ruleid: devsecops-framework.secure-coding.java.trust-all-certificates
    public void checkClientTrusted(X509Certificate[] chain, String authType) { }
}

class TrustManagerSafeSamples {
    // The pattern requires an EMPTY body. A manager that actually validates has
    // a body and is correctly not flagged.
    // ok: devsecops-framework.secure-coding.java.trust-all-certificates
    public void checkServerTrusted(X509Certificate[] chain, String authType)
            throws CertificateException {
        if (chain == null || chain.length == 0) {
            throw new CertificateException("empty chain");
        }
    }
}

// Real Spring Security configuration. The earlier fixture called an invented
// `csrfOf(http)` helper, which no rule targets and which therefore proved
// nothing -- again a defect in the fixture rather than in the rule.
class SecurityConfigSamples {
    public void configure(HttpSecurity http) throws Exception {
        // ruleid: devsecops-framework.secure-coding.java.csrf-disabled
        http.csrf().disable();
    }
}

class SecurityConfigLambdaSamples {
    public void configure(HttpSecurity http) throws Exception {
        // ruleid: devsecops-framework.secure-coding.java.csrf-disabled
        http.csrf(c -> c.disable());
    }
}

class SecurityConfigSafeSamples {
    public void configure(HttpSecurity http) throws Exception {
        // CSRF left enabled: configured, not disabled.
        // ok: devsecops-framework.secure-coding.java.csrf-disabled
        http.csrf();
    }
}

File: framework/rules/php/secure-coding.php
<?php
// Rule fixtures for framework/rules/php/secure-coding.yml
//
// Semgrep's own test convention: a `ruleid:` annotation asserts the rule MUST
// fire on the next line; `ok:` asserts it MUST NOT. `semgrep --test` reads both,
// so this file is simultaneously the positive test, the negative test, and the
// documentation of what each rule considers safe.
//
// The negative cases matter more than the positive ones. A rule that fires on
// the vulnerable line is easy; a rule that stays silent on the correct line is
// what determines whether developers leave it switched on.

function sql_injection_cases($conn, $safe_id) {
    // ruleid: devsecops-framework.secure-coding.php.sql-injection-superglobal
    mysqli_query($conn, "SELECT * FROM users WHERE id = $_GET[id]");

    // ok: devsecops-framework.secure-coding.php.sql-injection-superglobal
    $stmt = $conn->prepare("SELECT * FROM users WHERE id = ?");

    // ok: devsecops-framework.secure-coding.php.sql-injection-superglobal
    mysqli_query($conn, "SELECT * FROM users WHERE active = 1");
}

function sql_concatenation_cases($conn, $id) {
    // ruleid: devsecops-framework.secure-coding.php.sql-string-concatenation
    mysqli_query($conn, "SELECT * FROM users WHERE id = " . $id);

    // ok: devsecops-framework.secure-coding.php.sql-string-concatenation
    $stmt = $conn->prepare("SELECT * FROM users WHERE id = ?");
}

function xss_cases($safe_value) {
    // ruleid: devsecops-framework.secure-coding.php.xss-superglobal-echo
    echo $_GET['name'];

    // ok: devsecops-framework.secure-coding.php.xss-superglobal-echo
    echo htmlspecialchars($_GET['name'], ENT_QUOTES, 'UTF-8');

    // ok: devsecops-framework.secure-coding.php.xss-superglobal-echo
    echo "a static string";
}

function command_injection_cases() {
    // ruleid: devsecops-framework.secure-coding.php.command-execution-superglobal
    system("ping -c 1 $_GET[host]");

    // ok: devsecops-framework.secure-coding.php.command-execution-superglobal
    system("ping -c 1 " . escapeshellarg('127.0.0.1'));
}

function file_inclusion_cases() {
    // ruleid: devsecops-framework.secure-coding.php.file-inclusion-superglobal
    include $_GET['page'];

    // ok: devsecops-framework.secure-coding.php.file-inclusion-superglobal
    include 'templates/header.php';
}

function path_traversal_cases() {
    // ruleid: devsecops-framework.secure-coding.php.path-traversal-superglobal
    readfile($_GET['file']);

    // ok: devsecops-framework.secure-coding.php.path-traversal-superglobal
    readfile('/var/www/static/logo.png');
}

function dynamic_execution_cases() {
    // ruleid: devsecops-framework.secure-coding.php.dynamic-code-execution
    eval($_GET['code']);

    // ok: devsecops-framework.secure-coding.php.dynamic-code-execution
    $dispatch = ['list' => 'listAction'];
}

function deserialization_cases() {
    // ruleid: devsecops-framework.secure-coding.php.unserialize-superglobal
    $obj = unserialize($_COOKIE['prefs']);

    // ok: devsecops-framework.secure-coding.php.unserialize-superglobal
    $obj = json_decode($_COOKIE['prefs'], true);
}

function exception_disclosure_cases($e) {
    try {
        risky();
    } catch (Exception $E) {
        // ruleid: devsecops-framework.secure-coding.php.exception-message-echoed
        echo $E->getMessage();
    }

    try {
        risky();
    } catch (Exception $E) {
        // ok: devsecops-framework.secure-coding.php.exception-message-echoed
        error_log($E->getMessage());
    }
}

function password_hash_cases($password, $file_contents) {
    // ruleid: devsecops-framework.secure-coding.php.weak-password-hash
    $hash = md5($password);

    // ok: devsecops-framework.secure-coding.php.weak-password-hash
    $hash = password_hash($password, PASSWORD_DEFAULT);

    // A checksum over file content is a legitimate use of md5 and must not fire.
    // ok: devsecops-framework.secure-coding.php.weak-password-hash
    $checksum = md5($file_contents);
}

function tls_cases($ch) {
    // ruleid: devsecops-framework.secure-coding.php.tls-verification-disabled
    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);

    // ok: devsecops-framework.secure-coding.php.tls-verification-disabled
    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, true);
}
